mcpserver/server.go

```diff
@@ -0,0 +1,465 @@
+// Copyright (c) 2023-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package mcpserver
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/mark3labs/mcp-go/server"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+)
+
+// MattermostMCPServer provides a high-level interface for creating an MCP server
+// with Mattermost-specific tools and authentication
+type MattermostMCPServer struct {
+	mcpServer    *server.MCPServer
+	authProvider AuthenticationProvider
+	logger       mlog.LoggerIFace
+	config       Config
+}
+
+// NewMattermostMCPServer creates a new Mattermost MCP server with the specified configuration
+func NewMattermostMCPServer(config Config, authProvider AuthenticationProvider, logger mlog.LoggerIFace) (*MattermostMCPServer, error) {
+	// Create the mcp-go server directly
+	mcpServer := server.NewMCPServer(
+		"mattermost-mcp-server",
+		"0.1.0",
+		server.WithToolCapabilities(true), // Enable tool list changed notifications
+		server.WithLogging(),              // Enable logging capabilities
+	)
+
+	mattermostServer := &MattermostMCPServer{
+		mcpServer:    mcpServer,
+		authProvider: authProvider,
+		logger:       logger,
+		config:       config,
+	}
+
+	// For standalone mode (stdio with PAT), validate token at startup
+	if config.Transport == "stdio" {
+		ctx := context.Background()
+		_, err := authProvider.ValidateAuth(ctx, config.PersonalAccessToken)
+		if err != nil {
+			return nil, fmt.Errorf("startup token validation failed: %w", err)
+		}
+	}
+
+	// Register all Mattermost tools
+	mattermostServer.registerMattermostTools()
+
+	return mattermostServer, nil
+}
+
+// Serve starts the server using the configured transport
+func (s *MattermostMCPServer) Serve() error {
+	switch s.config.Transport {
+	case "stdio", "": // default to stdio for backward compatibility
+		return s.serveStdio()
+	case "http":
+		return s.serveHTTP()
+	default:
+		return fmt.Errorf("unsupported transport type: %s", s.config.Transport)
+	}
+}
+
+// serveStdio starts the server using stdio transport
+func (s *MattermostMCPServer) serveStdio() error {
+	// Configure error logger to use our mlog logger if available
+	var errorLogger *log.Logger
+	if s.logger != nil {
```

Comment:
We can probably assume we will always pass this. Maybe we just accept the standard log interface to the constructor?